src/main/java/com/okta/tools/helpers/HttpHelper.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2018 Okta
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.tools.helpers;
+
+import org.apache.http.HttpHost;
+import org.apache.http.config.Registry;
+import org.apache.http.config.RegistryBuilder;
+import org.apache.http.conn.socket.ConnectionSocketFactory;
+import org.apache.http.conn.socket.PlainConnectionSocketFactory;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
+import org.apache.http.protocol.HttpContext;
+import org.apache.http.protocol.HttpCoreContext;
+import org.apache.http.ssl.SSLContexts;
+
+import javax.net.ssl.SSLContext;
+import java.io.IOException;
+import java.net.*;
+
+// Inspired by https://stackoverflow.com/a/22960881/154527
+public final class HttpHelper {
+    public static CloseableHttpClient createClient(HttpClientBuilder httpClientBuilder)
+    {
+        Registry<ConnectionSocketFactory> reg = RegistryBuilder.<ConnectionSocketFactory>create()
+                .register("http", ProxySelectorPlainConnectionSocketFactory.INSTANCE)
+                .register("https", new ProxySelectorSSLConnectionSocketFactory(SSLContexts.createSystemDefault()))
+                .build();
+        PoolingHttpClientConnectionManager cm = new PoolingHttpClientConnectionManager(reg);
+        return httpClientBuilder
+                .useSystemProperties()
+                .setConnectionManager(cm)
+                .build();
+    }
+
+    public static CloseableHttpClient createClient()
+    {
+        return createClient(HttpClients.custom());
+    }
+
+    private enum ProxySelectorPlainConnectionSocketFactory implements ConnectionSocketFactory {
+        INSTANCE;
+
+        @Override
+        public Socket createSocket(HttpContext context) throws IOException {
+            Socket socket = HttpHelper.createSocket(context);
+            return socket != null ? socket : PlainConnectionSocketFactory.INSTANCE.createSocket(context);
```

Comment:
Attempt to create a compatible proxy socket adapter. Fall back to default plain connection implementation if a compatible proxy socket adapter cannot be created.